data/trie/trieStorageManager.go

```diff
@@ -177,45 +184,66 @@ func (tsm *trieStorageManager) Prune(rootHash []byte, identifier data.TriePrunin
 	tsm.storageOperationMutex.Lock()
 	defer tsm.storageOperationMutex.Unlock()
 
-	log.Trace("trie storage manager prune", "root", rootHash)
 	rootHash = append(rootHash, byte(identifier))
 
 	if tsm.snapshotInProgress > 0 {
-		// TODO refactor pruning mechanism so that pruning will be done on rollback
-		// even if there is a snapshot in progress
+		rootHash = append(rootHash, byte(prune))
+		tsm.pruningBuffer.add(rootHash)
 
-		//TODO activate pruning when snapshot is in progress
 		return
 	}
 
 	oldHashes := tsm.pruningBuffer.removeAll()
-	oldHashes[string(rootHash)] = struct{}{}
-	tsm.prune(oldHashes)
+	tsm.resolveBufferedHashes(oldHashes)
+	tsm.prune(rootHash)
 }
 
-func (tsm *trieStorageManager) prune(oldHashes map[string]struct{}) {
-	for hash := range oldHashes {
-		rootHash := []byte(hash)
-		err := tsm.removeFromDb(rootHash)
-		if err != nil {
-			log.Error("trie storage manager remove from db", "error", err, "rootHash", hex.EncodeToString(rootHash))
+func (tsm *trieStorageManager) resolveBufferedHashes(oldHashes [][]byte) {
+	for _, rootHash := range oldHashes {
+		lastBytePos := len(rootHash) - 1
```

Comment:
Maybe you should add a check for lastBytePos to be >= 0, otherwise the next code could panic

Reply:
:+1:

Reply:
Done.